pulp_rpm/app/downloaders.py

```diff
@@ -94,3 +106,115 @@ async def _run(self, extra_data=None):
+    """
+
+    def __init__(self, *args, uln_user=None, uln_pwd=None, uln_server_base_url=None, **kwargs):
+        """
+        Initialize the downloader for ULN repositories.
+
+        If no server URL is given, use the server url from oracle linux.
+        """
+        self.uln_user = uln_user
+        self.uln_pwd = uln_pwd
+        if uln_server_base_url:
+            self.uln_server_base_url = uln_server_base_url
+        else:  # if no uln_server_base_url is given, use oracle linux url
+            self.uln_server_base_url = "https://linux-update.oracle.com/"
+        self.headers = None
+        self.session_key = None
+
+        super().__init__(*args, **kwargs)
+
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        Once per session the coroutine logs into the ULN account using the
+        uln_username and uln_pwd. The returned key is used for authentification
+        for all other downloads.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+        """
+        parsed = urlparse(self.url)
+        if parsed.scheme == "uln":
+            if not (self.uln_user and self.uln_pwd):
+                raise NoUlnCredentialsError()
+            # get ULN Session-key
+            SERVER_URL = os.path.join(self.uln_server_base_url, "rpc/api")
+
+            # set proxy for authentification
+            client = my_ServerProxy(SERVER_URL, proxy=self.proxy, auth=self.auth)
+            if not self.session_key:
+                self.session_key = await client["auth.login"](self.uln_user, self.uln_pwd)
+                if len(self.session_key) != 43:
+                    raise NoUlnCredentialsError()
+                self.headers = {"X-ULN-API-User-Key": self.session_key}
+                await client.close()
+            # build request url from input uri
+            channelLabel = parsed.netloc
+            path = parsed.path.lstrip("/")
+            url = os.path.join(self.uln_server_base_url, "XMLRPC/GET-REQ", channelLabel, path)
+
+        async with self.session.get(
+            url, proxy=self.proxy, auth=self.auth, headers=self.headers
+        ) as response:
+            self.raise_for_status(response)
+            to_return = await self._handle_response(response)
+            await response.release()
+            self.response_headers = response.headers
+
+        if self._close_session_on_finalize:
+            self.session.close()
+            client.close()
+        return to_return
+
+
+# until aiohttp-xmlrpc allows http post with proxy, use the following patch
+# this only works for http connection to the proxy, https connections are not supported!
+class my_ServerProxy(ServerProxy):
+    """
+    Overwriting the class aiohttp_xmlrpc.ServreProxy to allow proxy handling.
+    """
+
+    def __init__(self, *args, proxy, auth, **kwargs):
```

Comment:
I believe `proxy, auth` do not need to be explicitly stated in the signature here.
Is the goal to make them required positional arguments?

Reply:
I'm not sure if I can leave them in the kwargs as I want to set them as a class argument. In the aioxmlrpc call in UlnDownloader I don't know how to set these parameters otherwise... Maybe you know a better or more elegant way to solve this?